GSoC 2023: JSR-305 probe blog (#6653)
jenkins-infra/jenkins.io: content/blog/2023/09/03/2023-09-03-jsr305-probe-blog.adoc

---
layout: post
title: JSR-305 Probe Blog
tags:
- gsoc
- gsoc2023
- healthscore
- probes
- plugin
authors:
- jagruti
opengraph:
  image: /images/gsoc/2023/jagruti/jsr-305-probe-image.png
sig:
- gsoc
---

== Introduction

> `@Nonnull` and `@CheckForNull` annotations (part of JSR-305) have been link:https://www.jenkins.io/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016]. For security reasons it is important that the outdated annotations are highlighted and replaced with `javax.annotation.Nonnull` and `edu.umd.cs.findbugs.annotations.NonNull`

Reviewer: "For security reasons" -- that sounds misleading. The real reasons are mentioned in https://groups.google.com/g/jenkinsci-dev/c/uE1wwtVi1W0/m/gLxdEJmlBQAJ.
Also `javax.annotation.Nonnull` is one of the outdated annotations, nothing should be replaced with it.

Author: I have added more information. Thanks for reviewing.

---
layout: post
title: Detecting deprecated JSR-305 imports Probe
tags:
- gsoc
- gsoc2023
- healthscore
- probes
- plugin
authors:
- jagruti
opengraph:
  image: /images/gsoc/2023/jagruti/jsr-305-probe-image.png
sig:
- gsoc
---

== Introduction

The `JSR-305` framework provides annotations to detect software defects. Its executive committee voted it as dormant since May 2012 (see: link:https://jcp.org/en/jsr/detail?id=305[JSR 305: Annotations for Software Defect Detection
])

> 

Author: In the above link: https://jcp.org/en/jsr/detail?id=305 the reason is mentioned:

> Reason: The Executive Committee voted to list this JSR as dormant in May 2012

It is important to stop using the framework for the following reasons:

- It is not maintained anymore.
- The annotations have questionable license.
- The annotations are in the reserved javax namespace and there is no public release of the spec (Nor is there ever likely to be. See point 1).
(Source: link:https://groups.google.com/g/jenkinsci-dev/c/uE1wwtVi1W0/m/gLxdEJmlBQAJ[Jenkins Developers Discussion])

`@Nonnull` and `@CheckForNull` annotations (part of JSR-305) have been link:https://www.jenkins.io/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016].

The `JSR-305` probe checks whether the deprecated annotations are used in a plugin.

== Importance of the probe

> This probe will identify the outdated annotations used in the plugin and highlight them in the scoring system.

Author: The scoring system hasn't been implemented yet. I do not have more information about it.

> == Challenges and Learning
> 
> The outdated annotations are imported in `.java` files. My first approach was to go for an abstract class because I assumed there could be other probes that may require to look for code in `.java` files.
> After a series of reviews, one of the mentors pointed out that to comply with the link:https://github.com/openrewrite/rewrite-jenkins/blob/98bfbecc9c9507a069c9ac1f1e35cd5dd0fa0687/src/main/resources/META-INF/rewrite/rewrite.yml[OpenWrite rule] all imports from the `javax.annotation` package should be checked for deprecation.
> 
> Also, based on the link:https://blog.devgenius.io/key-principles-in-software-dry-kiss-yagni-solid-and-other-acronyms-98e5575a6942[YAGNI principle] ("You Aren't Gonna Need It") which means develop features only when you need it, the `abstract` wasn't the current requirement. I restructured the code and used a single class to implement the probe.

Reviewer 2: ```suggestion
== Challenges and Learning

The outdated annotations are imported in `.java` files.
My first approach was to go for an abstract class because I assumed there could be other probes that may need to look for code in `.java` files.
After a series of reviews, one of the mentors pointed out that to comply with the link:https://github.com/openrewrite/rewrite-jenkins/blob/98bfbecc9c9507a069c9ac1f1e35cd5dd0fa0687/src/main/resources/META-INF/rewrite/rewrite.yml[OpenWrite rule], all imports from the `javax.annotation` package should be checked for deprecation.

Also, based on the link:https://blog.devgenius.io/key-principles-in-software-dry-kiss-yagni-solid-and-other-acronyms-98e5575a6942[YAGNI principle] ("You Aren't Gonna Need It"), which means features are developed only when you need them, the `abstract` wasn't the current requirement.
I restructured the code and used a single class to implement the probe.
```

suggestion for formatting and some syntax

Were there any additional challenges in restructuring the code or using a single class?

Author: There weren't any other challenges. I am just sharing the thought process of arriving at the solution here to help the future GSoCers/contributors.

> == Conclusion and Outcome
> This probe will identify the deprecated annotations and let the maintainers know which imports in a plugin require replacement.

Reviewer 2: ```suggestion
== Conclusion and Outcome
This probe will identify the deprecated annotations and let maintainers know which imports in a plugin require replacement.
```

Suggestion for syntax

---
layout: post
title: JSR-305 Probe Blog
tags:
- gsoc
- gsoc2023
- healthscore
- probes
- plugin
authors:
- jagruti
opengraph:
  image: /images/gsoc/2023/jagruti/jsr-305-probe-image.png
sig:
- gsoc
---

== Introduction

The `JSR-305` framework provides annotations to detect software defects. Its executive committee voted it as dormant since May 2012 (see: link:https://jcp.org/en/jsr/detail?id=305[JSR 305: Annotations for Software Defect Detection
])

It is important to stop using the framework for the following reasons:

- It is not maintained anymore.
- The annotations have questionable license.
- The annotations are in the reserved javax namespace and there is no public release of the spec (Nor is there ever likely to be. See point 1).
(Source: link:https://groups.google.com/g/jenkinsci-dev/c/uE1wwtVi1W0/m/gLxdEJmlBQAJ[Jenkins Developers Discussion])

`@Nonnull` and `@CheckForNull` annotations (part of JSR-305) have been link:https://www.jenkins.io/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016].

> The `javax.annotation.Nonnull` import should be replaced with `edu.umd.cs.findbugs.annotations.NonNull`.

Reviewer 3: Why giving this example and not the replacement for `CheckForNull`? Not sure it is very important to have here?

---
layout: post
title: Detecting deprecated JSR-305 imports Probe
tags:
- gsoc
- gsoc2023
- healthscore
- probes
- plugin
authors:
- jagruti
opengraph:
  image: /images/gsoc/2023/jagruti/jsr-305-probe-image.png
sig:
- gsoc
---

== Introduction

> This blog post describes the latest addition to the plugin health scoring system developed as part of link:../../../../projects/gsoc/2023/projects/add-probes-to-plugin-health-score.adoc[my Google summer of Code Project]: a probe that detects usages of the `JSR-305` framework in plugins.

Reviewer 2: ```suggestion
This blog post describes the latest addition to the plugin health scoring system developed as part of link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[my Google summer of Code Project]: a probe that detects usages of the `JSR-305` framework in plugins.
```
Same as other link suggestions

The link:https://jcp.org/en/jsr/detail?id=305[`JSR-305` framework] provides annotations to detect software defects.
Its executive committee link:https://jcp.org/en/jsr/detail?id=305[voted it as dormant since May 2012].
It is important to stop using the framework for the following reasons:

- It is no longer maintained.
- The annotations have questionable license.
- The annotations are in the reserved javax namespace and there is no public release of the spec.
Since it is not actively maintained, there is little chance this would be published.

For more insight, refer to the link:https://groups.google.com/g/jenkinsci-dev/c/uE1wwtVi1W0/m/gLxdEJmlBQAJ[Jenkins developer mailing list conversation].

> The `@Nonnull` and `@CheckForNull` annotations, which are part of JSR-305, have been link:https://www.jenkins.io/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016].

Reviewer 2: ```suggestion
The `@Nonnull` and `@CheckForNull` annotations, which are part of JSR-305, have been link:/doc/developer/tutorial-improve/replace-jsr-305-annotations/[deprecated since 2016].
```
Suggestion to remove the extraneous part of this link

This probe identifies the outdated annotations used in a plugin and highlight them in the scoring system.

== Challenges and Learning

The outdated annotations are imported in `.java` files.
My first approach was to go for an abstract class because I assumed there could be other probes that may need to look for code in `.java` files.
After a series of reviews, one of the mentors pointed out that to comply with the link:https://github.com/openrewrite/rewrite-jenkins/blob/98bfbecc9c9507a069c9ac1f1e35cd5dd0fa0687/src/main/resources/META-INF/rewrite/rewrite.yml[OpenWrite rule], all imports from the `javax.annotation` package should be checked for deprecation.

Also, based on the link:https://blog.devgenius.io/key-principles-in-software-dry-kiss-yagni-solid-and-other-acronyms-98e5575a6942[YAGNI principle] ("You Aren't Gonna Need It"), which means features are developed only when you need them, the `abstract` wasn't the current requirement.
I restructured the code and used a single class to implement the probe.

== Conclusion and Outcome

This probe will identify the deprecated annotations and let maintainers know which imports in a plugin require replacement.

== Links

> For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:../../../../projects/gsoc/2023/projects/add-probes-to-plugin-health-score.adoc[Adding Probes to "Plugin Health Score"] description page.

Reviewer 2: ```suggestion
For more information or to find answers to any questions you might have, please visit the official GSoC 2023 project link:/projects/gsoc/2023/projects/add-probes-to-plugin-health-score/[Adding Probes to "Plugin Health Score"] description page.
```
Same as other link suggestions
